pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,293 @@
+)
+
+// machineDeletionHookName holds a name of the Machine Deletion Hook
+const machineDeletionHookName = "EtcdQuorumOperator"
+
+// machineDeletionHookOwner holds an owner of the Machine Deletion Hook
+const machineDeletionHookOwner = "clusteroperator/etcd"
+
+type machineDeletionHooksController struct {
+	etcdClient                            etcdcli.EtcdClient
+	machineClient                         machinev1beta1client.MachineInterface
+	masterMachineLister                   machinelistersv1beta1.MachineLister
+	configMapListerForKubeSystemNamespace corev1listers.ConfigMapNamespaceLister
+	// machineAPIChecker determines if the precondition for this controller is met,
+	// this controller can be run only on a cluster that exposes a functional Machine API
+	machineAPIChecker     ceohelpers.MachineAPIChecker
+	masterMachineSelector labels.Selector
+}
+
+// NewMachineDeletionHooksController reconciles machine hooks for master machines
+//
+// note:
+//  The Machine Deletion Hook this controller reconciles is a mechanism within the Machine API that allow this operator to hold up removal of a machine
+//  until a replacement member has been promoted to a voting member.
+func NewMachineDeletionHooksController(
+	operatorClient operatorv1helpers.OperatorClient,
+	machineClient machinev1beta1client.MachineInterface,
+	etcdClient etcdcli.EtcdClient,
+	machineAPIChecker ceohelpers.MachineAPIChecker,
+	masterMachineSelector labels.Selector,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	masterMachineInformer cache.SharedIndexInformer,
+	eventRecorder events.Recorder) factory.Controller {
+	c := &machineDeletionHooksController{
+		machineClient:                         machineClient,
+		etcdClient:                            etcdClient,
+		machineAPIChecker:                     machineAPIChecker,
+		masterMachineSelector:                 masterMachineSelector,
+		masterMachineLister:                   machinelistersv1beta1.NewMachineLister(masterMachineInformer.GetIndexer()),
+		configMapListerForKubeSystemNamespace: kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Lister().ConfigMaps("kube-system"),
+	}
+	return factory.New().
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		WithInformers(
+			masterMachineInformer,
+			kubeInformersForNamespaces.InformersFor("kube-system").Core().V1().ConfigMaps().Informer(),
+		).
+		ResyncEvery(1*time.Minute). // an arbitrary number, not too high not too low
+		ToController("MachineDeletionHooksController", eventRecorder.WithComponentSuffix("cluster-machine-deletion-hooks-controller"))
+}
+
+func (c *machineDeletionHooksController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	// stop if the machine API is not functional
+	if isFunctional, err := c.machineAPIChecker.IsFunctional(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	if err := c.addDeletionHookToMasterMachines(ctx); err != nil {
```

Comment:
an error here will stop you from removing your machine deletion hook.  you sure you want to combine them?

Reply:
updated